Introduce Search API

stephenplusplus opened the pull request:
Fixes #632 

Docs: https://stephenplusplus.github.io/gcloud-node/#/docs/master/search

To dos:
- [x] Enable stream mode in `Search#getIndexes()`
- [x] Enable stream mode in `Index#getDocuments()`
- [x] Enable stream mode in `Index#search()`
- [x] Docs
  - [x] JSDocs
  - [x] Readme
  - [x] Site
    - [x] Update sitemap
    - [x] Add "Search Overview" section
  - [x] StreamRouter
- [x] Tests
  - [x] Unit
  - [x] System

API can't-do's:
- Can't delete an index (can list indexes and delete all documents)
- Can't create an index directly (only when creating a document)
- Can't search a specific document

Not perfected yet:
- ~~Document JSON (when creating a document, we may be able to ease the field specification process)~~
  - See example below
- ~~Queries (we can probably have an object builder rather than just accepting a string)~~
  - It seems better to allow the developer to use the familiar query language as outlined in [the docs](https://cloud.google.com/search/query). Our re-implementing the grammar could introduce inconsistencies and prevent portability.

Examples:
#### Create an index / Create a document from scratch

``` js
var search = gcloud.search();
var index = search.index('gcloud-services');

var document = index.document('search');
document.addField('implemented').addAtomValue('almost');

index.createDocument(document, function(err, document, apiResponse) {});
```
#### Create a document from an existing JSON object / file

``` js
var search = gcloud.search();
var index = search.index('gcloud-services');

var documentJson = {
  docId: 'search',
  fields: {
    values: [
      implemented: {
        stringValue: 'almost',
        stringFormat: 'ATOM'
      }
    ]
  }
};

index.createDocument(documentJson, function(err, document, apiResponse) {});
```
#### Get search results with a callback

``` js
var search = gcloud.search();
var index = search.index('gcloud-services');

function onSearchResults(err, documents, nextQuery, apiResponse) {
  if (err) {
    console.error(err);
    return;
  }

  // documents[0].id = 'search';

  if (nextQuery) {
    index.search(nextQuery, onSearchResults);
  }
}

index.search('implemented:almost', onSearchResults);
```
#### Stream search results

``` js
var search = gcloud.search();
var index = search.index('gcloud-services');

index.search('implemented:almost')
  .on('error', console.error)
  .on('data', function(document) {
    // document.id = 'search';
  });
```

stephenplusplus commented on system-test/search.js:
```diff
@@ -0,0 +1,182 @@
+    deleteAllDocuments(done);
+  });
+
+  it('should create a document in a new index', function(done) {
```
Example usage.

jgeewax commented on lib/search/document.js:
```diff
@@ -0,0 +1,69 @@
+};
+
+Document.prototype.makeReq_ = function(method, path, query, body, callback) {
+  path = '/documents/' + this.id + path;
```
The path to the document is /indexes/{indexId}/documents/{documentId} right?

stephenplusplus replied:
Yeah, we are re-using the `makeReq_` from the index, so the index components are already in place.

jgeewax commented on lib/search/index-class.js:
```diff
@@ -0,0 +1,140 @@
+  this.id = id;
+}
+
+Index.prototype.createDocument = function(document, callback) {
```
Do we have to worry about `document` being a reserved word in JS?

stephenplusplus replied:
Only if we do a browser build. In which case, there are a lot of variables there; we might be able to rename it at the time of export (to document_ or something). It's something we can figure out quite easily when that time comes.

jgeewax commented on lib/search/index-class.js:
```diff
@@ -0,0 +1,140 @@
+  return new Document(this, id);
+};
+
+Index.prototype.getDocuments = function(query, callback) {
```
Does this follow the `on('data', ...)` pattern as well?

stephenplusplus replied:
It can! (will make a note)

stephenplusplus commented:
`Index#getDocuments` and `Index#search` are now stream-able.

Also, I added document building functionality and updated the first post here with some examples.

stephenplusplus commented:
The todo list is a-ticking away. Now would be a good time to catch any mis-steps (see [the docs](https://stephenplusplus.github.io/gcloud-node/#/docs/search)). Next to tackle are the tests.

stephenplusplus commented:
This is ready for review (from who, I don't know :smile:). Todos completed.

ryanseys commented:
I will
On Jun 17, 2015 6:13 AM, "Stephen Sawchuk" <notifications@github.com> wrote:

> This is ready for review (from who, I don't know [image: :smile:]). Todos
> completed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-node/pull/648#issuecomment-112794551>
> .
>

ryanseys commented on lib/common/stream-router.js:
```diff
@@ -0,0 +1,118 @@
+
+/*! Developer Documentation
+ *
+ * streamRouter is used to extend `nextQuery`+callback methods with stream
```
:+1:

ryanseys commented on lib/index.js:
```diff
@@ -195,6 +203,33 @@ gcloud.pubsub = function(config) {
+ *
+ * @example
+ * var gcloud = require('gcloud');
+ * var seach = gcloud.search({
```
`var search`

ryanseys commented on lib/search/index-class.js:
```diff
@@ -0,0 +1,385 @@
+ * Get {module:search/document} objects for all of the documents in this index.
+ *
+ * @param {object=} query - Query object.
+ * @param {string} query.pageSize - The maximum number of indexes to return per
```
I think you mean to say the maximum number of documents to return per page.

ryanseys commented on lib/search/index-class.js:
```diff
@@ -0,0 +1,385 @@
+ * @throws {error} If a query string or object isn't provided.
+ *
+ * @param {string|object} query - A query object or simply a string query.
+ * @param {string} query.pageSize - The maximum number of indexes to return per
```
Again, I think you mean documents here, not indexes.

ryanseys commented on lib/search/index.js:
```diff
@@ -0,0 +1,232 @@
+ * Get {module:search/index} objects for all of the indexes in your project.
+ *
+ * @param {object=} query - Query object.
+ * @param {string} query.pageSize - The maximum number of indexes to return per
```
Now here, you really do mean indexes :)

ryanseys commented on lib/search/index.js:
```diff
@@ -0,0 +1,232 @@
+ * @type {module:search/index}
+ * @private
+ */
+var Index = require('./index-class.js');
```
haha. That naming.

ryanseys commented on lib/search/document.js:
```diff
@@ -0,0 +1,213 @@
+ * var scoreField = document.addField('score');
+ * // scoreField is a Field object.
+ */
+Document.prototype.addField = function(name) {
```
So the idea here is that you add the fields to the document, and then call Index#createDocument with the document object and the fields will be included? Yes? That's unclear here, so perhaps is there some way to improve the documentation in that regard?

stephenplusplus replied:
Yeah, we usually document just the behavior of the method in its docs and use the readme to hold the "here's how you piece it all together" docs.

This probably won't happen in the immediate future, but #666 will help with this.

Maybe for now, an example in the `Document` constructor would be helpful to show the creation process.

ryanseys commented:
I run `npm test` and get the following console logs at the end of the test results (no failures):

```
{ [Error: ENOENT, open '/path/to/keyfile.json'] errno: -2, code: 'ENOENT', path: '/path/to/keyfile.json' }
{ [Error: ENOENT, open '/path/to/keyfile.json'] errno: -2, code: 'ENOENT', path: '/path/to/keyfile.json' }
{ [Error: ENOENT, open '/path/to/keyfile.json'] errno: -2, code: 'ENOENT', path: '/path/to/keyfile.json' }
```

stephenplusplus commented:
I thought that might just be me. Can you advise how to debug that?

ryanseys commented:
I think it's because we pass the external `console` object into the `vm` object when we test our code docs, so if the docs that are tested have any calls to `console.log` or `console.error` then they will be written to our external stdout. This is the case with some of your tests when you run `.on('error', console.error)`.

Easiest way to fix is overwrite `console.log` and `console.error` for the docs test.

ryanseys commented on system-test/search.js:
```diff
@@ -0,0 +1,301 @@
+  });
+
+  describe('search', function() {
+    var query = 'ryan';
```
yay!

ryanseys commented on system-test/search.js:
```diff
@@ -0,0 +1,301 @@
+
+      var questions = document.addField('question');
+      questions.addTextValue('Where did Ryan go?');
+      questions.addTextValue('Where did Silvano go?');
```
:(

ryanseys replied:
@silvolu

ryanseys commented on test/docs.js:
```diff
@@ -78,6 +78,7 @@ describe('documentation', function() {
         gcloud: gcloud,
         require: require,
         process: process,
+        console: console,
```
We'll have to overwrite these methods if we don't want the docs test to explode in our terminal when run. :)

stephenplusplus replied:
Thanks for explaining this :)

stephenplusplus commented on test/docs.js:
```diff
@@ -74,10 +74,16 @@ describe('documentation', function() {
         ].join('\n'));
       }
 
+      var mockConsole = Object.keys(console).reduce(function(console, method) {
```
@ryanseys this look ok?

ryanseys replied:
:+1:

ryanseys commented on README.md:
```diff
@@ -273,8 +317,12 @@ Apache 2.0 - See [COPYING](COPYING) for more information.
 [cloud-pubsub-docs]: https://cloud.google.com/pubsub/docs
 
+[cloud-search]: https://cloud.google.com/search/
+[cloud-search-docs]: https://cloud.google.com/search/
```
Nitpick: These two links are the same.

stephenplusplus replied:
Yeah :) None of the docs are fully public yet, and there isn't yet a landing page like the others.

ryanseys replied:
Ah yeah.

ryanseys commented on docs/site/components/docs/search-overview.html:
```diff
@@ -0,0 +1,7 @@
+<h3>Search Overview</h3>
+<p>
+  The object returned from <code>gcloud.search</code> gives you complete access to store and search your indexes and documents.
```
Nitpick: What do we search? Do we search an index or search a document?

stephenplusplus replied:
Changed to "gives you complete access to store your documents and search your indexes."

ryanseys replied:
:+1:

ryanseys commented:
Looks GREAT to me!

stephenplusplus commented on lib/search/document.js:
```diff
@@ -0,0 +1,231 @@
+ *
+ * var document = search.index('records').document('stephen');
+ *
+ * //-
```
@ryanseys added these docs re: https://github.com/GoogleCloudPlatform/gcloud-node/pull/648#discussion_r32641544 but it feels a little out of place. I could go either way though. LMKWYT.

ryanseys replied:
Yeah, seems out of place because its not necessary in this place. But it is necessary to createDocument after calling addField (does that make sense?) so maybe the docs should be added there or a note should be made on that method instead.

stephenplusplus replied:
Not totally following. I'll remove this example. Note the docs for [createDocument](https://stephenplusplus.github.io/gcloud-node/#/docs/master/search/index?method=createDocument) show the workflow pretty well.

stephenplusplus commented:
Wee! I just added this in to handle pre-mature closing of a stream: https://github.com/stephenplusplus/gcloud-node/commit/abca6f186c2046fac99bbcfa2459822a5f3832f3

Basically to handle something like this:

```js
index.getDocuments().on('data', function(document) {
  if (document.id === 'the-one-i-want') {
    this.end();
  }
});
```

Before the change, we would have had no idea `end` was called, and would have kept trying to push documents to the stream & even continue calling the API for more results.

stephenplusplus commented:
Merge-time!
